File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ScannerCoordinator.kt
package expo.modules.vescapecore.connection

import expo.modules.vescapecore.diagnostics.ConnectionTraceReason
import java.util.concurrent.atomic.AtomicLong

/**
 * Why the BLE scanner is running. One vocabulary for every scan in the app, so a callback can say
 * which operation it belongs to instead of the code guessing from surrounding state.
 *
 * @parity /modules/vescape-core/ios/connection/ScannerCoordinator.swift `ScanPurpose`
 * @parity /modules/vescape-core/src/index.ts `ScanPurpose`
 */
enum class ScanPurpose(val wireValue: String, val owner: ConnectionOwner) {
    /** Foreground-entry Board Presence Scan (ADR 0035). Yields to every exclusive owner. */
    Presence("presence", ConnectionOwner.AutoConnect),

    /** Rider-driven Add Board discovery. Exclusive: never preempted. */
    AddBoard("add_board", ConnectionOwner.AddBoardScan),

    /** Board Probe handshake during linking/setup. Exclusive: never preempted. */
    BoardProbe("board_probe", ConnectionOwner.BoardProbe),

    /** Search backing an explicit Connect Intent. */
    ConnectIntent("connect_intent", ConnectionOwner.ConnectIntent),

    /** Mid-ride rediscovery of the Board Session's own peripheral. */
    Reconnect("reconnect", ConnectionOwner.BoardSession),
    ;

    val isExclusive: Boolean get() = owner.isExclusiveScannerOwner

    companion object {
        fun fromWire(value: String?): ScanPurpose? = entries.firstOrNull { it.wireValue == value }
    }
}

/** A granted scan, identified by its operation token. */
data class ScanOperation(val token: Long, val purpose: ScanPurpose)

/** Result of asking [ScannerCoordinator] for the radio. */
sealed interface ScanAcquisition {
    data class Granted(val operation: ScanOperation) : ScanAcquisition

    data class Denied(val reason: String, val heldBy: ScanPurpose?) : ScanAcquisition
}

/**
 * The one arbiter of BLE scanner ownership. Every scan takes a token; BLE scan callbacks outlive
 * their operation, so a callback that cannot prove it is [isCurrent] is dropped rather than allowed
 * to mutate state belonging to a newer scan.
 *
 * Add Board scan and Board Probe hold the scanner exclusively and cannot be preempted — a Presence
 * Scan asking while either runs is denied with `scanner_busy`.
 *
 * @parity /modules/vescape-core/ios/connection/ScannerCoordinator.swift
 */
class ScannerCoordinator {
    companion object {
        /** Process-wide arbiter. Add Board scan, Board Probe, reconnect, and Presence Scan share it. */
        val shared = ScannerCoordinator()
    }

    private val tokens = AtomicLong(0)

    @Volatile
    var active: ScanOperation? = null
        private set

    val activePurpose: ScanPurpose? get() = active?.purpose

    @Synchronized
    fun acquire(purpose: ScanPurpose): ScanAcquisition {
        val holder = active
        if (holder != null && holder.purpose != purpose && !purpose.owner.outranks(holder.purpose.owner)) {
            return ScanAcquisition.Denied(ConnectionTraceReason.SCANNER_BUSY, holder.purpose)
        }
        val operation = ScanOperation(token = tokens.incrementAndGet(), purpose = purpose)
        active = operation
        return ScanAcquisition.Granted(operation)
    }

    /** True only for the operation that still owns the scanner. Stale callbacks fail here. */
    fun isCurrent(operation: ScanOperation?): Boolean =
        operation != null && active?.token == operation.token

    /** Release only if [operation] still owns the scanner, so a late stop cannot kill a newer scan. */
    @Synchronized
    fun release(operation: ScanOperation?): Boolean {
        if (!isCurrent(operation)) return false
        active = null
        return true
    }
}

File: modules/vescape-core/ios/connection/ScannerCoordinator.swift
import Foundation

/// Why the BLE scanner is running. One vocabulary for every scan in the app, so a callback can say
/// which operation it belongs to instead of the code guessing from surrounding state.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ScannerCoordinator.kt `ScanPurpose`
/// @parity /modules/vescape-core/src/index.ts `ScanPurpose`
internal enum ScanPurpose: CaseIterable {
  /// Foreground-entry Board Presence Scan (ADR 0035). Yields to every exclusive owner.
  case presence
  /// Rider-driven Add Board discovery. Exclusive: never preempted.
  case addBoard
  /// Board Probe handshake during linking/setup. Exclusive: never preempted.
  case boardProbe
  /// Search backing an explicit Connect Intent.
  case connectIntent
  /// Mid-ride rediscovery of the Board Session's own peripheral.
  case reconnect

  var wireValue: String {
    switch self {
    case .presence: return "presence"
    case .addBoard: return "add_board"
    case .boardProbe: return "board_probe"
    case .connectIntent: return "connect_intent"
    case .reconnect: return "reconnect"
    }
  }

  var owner: ConnectionOwner {
    switch self {
    case .presence: return .autoConnect
    case .addBoard: return .addBoardScan
    case .boardProbe: return .boardProbe
    case .connectIntent: return .connectIntent
    case .reconnect: return .boardSession
    }
  }

  var isExclusive: Bool { owner.isExclusiveScannerOwner }

  static func fromWire(_ value: String?) -> ScanPurpose? {
    allCases.first { $0.wireValue == value }
  }
}

/// A granted scan, identified by its operation token.
internal struct ScanOperation: Equatable {
  let token: Int64
  let purpose: ScanPurpose
}

/// Result of asking `ScannerCoordinator` for the radio.
internal enum ScanAcquisition {
  case granted(ScanOperation)
  case denied(reason: String, heldBy: ScanPurpose?)
}

/// The one arbiter of BLE scanner ownership. Every scan takes a token; BLE scan callbacks outlive
/// their operation, so a callback that cannot prove it is `isCurrent` is dropped rather than allowed
/// to mutate state belonging to a newer scan.
///
/// Add Board scan and Board Probe hold the scanner exclusively and cannot be preempted — a Presence
/// Scan asking while either runs is denied with `scanner_busy`.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ScannerCoordinator.kt
internal final class ScannerCoordinator {
  /// Process-wide arbiter. Add Board scan, Board Probe, reconnect, and Presence Scan share it.
  static let shared = ScannerCoordinator()

  private let lock = NSLock()
  private var tokens: Int64 = 0
  private var operation: ScanOperation?

  var active: ScanOperation? {
    lock.lock()
    defer { lock.unlock() }
    return operation
  }

  var activePurpose: ScanPurpose? { active?.purpose }

  func acquire(_ purpose: ScanPurpose) -> ScanAcquisition {
    lock.lock()
    defer { lock.unlock() }
    if let holder = operation, holder.purpose != purpose, !purpose.owner.outranks(holder.purpose.owner) {
      return .denied(reason: ConnectionTraceReason.scannerBusy, heldBy: holder.purpose)
    }
    tokens += 1
    let granted = ScanOperation(token: tokens, purpose: purpose)
    operation = granted
    return .granted(granted)
  }

  /// True only for the operation that still owns the scanner. Stale callbacks fail here.
  func isCurrent(_ candidate: ScanOperation?) -> Bool {
    guard let candidate else { return false }
    return active?.token == candidate.token
  }

  /// Release only if `candidate` still owns the scanner, so a late stop cannot kill a newer scan.
  @discardableResult
  func release(_ candidate: ScanOperation?) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    guard let candidate, operation?.token == candidate.token else { return false }
    operation = nil
    return true
  }
}
